Adopt new AVPictureInPictureController for handling Picture In Picture

jyavenard opened the pull request:
#### 38735ed5707ade98e659cc3b5c033884d6235ba8
<pre>
Adopt new AVPictureInPictureController for handling Picture In Picture
<a href="https://bugs.webkit.org/show_bug.cgi?id=242061">https://bugs.webkit.org/show_bug.cgi?id=242061</a>
rdar://96066192

Reviewed by Dean Jackson.

For now, code is only enabled if UIKIT_WEBKIT_INTERNALS is set.
While the aim is to adopt this new API across all platforms,
it is currently only usable with UIKit.

Manually tested in the simulator.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerGetRawCookies const):
* Source/WebCore/platform/cocoa/VideoFullscreenInterfacePiP.h: Copied from Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h.
* Source/WebCore/platform/cocoa/VideoFullscreenInterfacePiP.mm: Added.
(clearUIColor):
(boolString):
(WebAVPictureInPictureContentViewController_initWithController):
(WebAVPictureInPictureContentViewController_controller):
(WebAVPictureInPictureContentViewController_playerLayer):
(WebAVPictureInPictureContentViewController_setPlayerLayer):
(WebAVPictureInPictureContentViewController_viewWillLayoutSubviews):
(WebAVPictureInPictureContentViewController_dealloc):
(allocWebAVPictureInPictureContentViewControllerInstance):
(-[WebAVPictureInPictureController fullscreenInterface]):
(-[WebAVPictureInPictureController setFullscreenInterface:]):
(-[WebAVPictureInPictureController initWithFullscreenInterface:]):
(-[WebAVPictureInPictureController dealloc]):
(-[WebAVPictureInPictureController tryToStartPictureInPicture]):
(-[WebAVPictureInPictureController stopPictureInPicture]):
(-[WebAVPictureInPictureController cleanup]):
(-[WebAVPictureInPictureController observeValueForKeyPath:ofObject:change:context:]):
(-[WebAVPictureInPictureController initObserver]):
(-[WebAVPictureInPictureController removeObserver]):
(-[WebAVPictureInPictureController pictureInPictureControllerWillStartPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureControllerDidStartPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureController:failedToStartPictureInPictureWithError:]):
(-[WebAVPictureInPictureController pictureInPictureControllerWillStopPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureControllerDidStopPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureController:restoreUserInterfaceForPictureInPictureStopWithCompletionHandler:]):
(WebAVPlayerLayerView_layerClass):
(WebAVPlayerLayerView_playerController):
(WebAVPlayerLayerView_setPlayerController):
(WebAVPlayerLayerView_playerLayer):
(allocWebAVPlayerLayerViewInstance):
(VideoFullscreenInterfacePiP::create):
(VideoFullscreenInterfacePiP::VideoFullscreenInterfacePiP):
(VideoFullscreenInterfacePiP::~VideoFullscreenInterfacePiP):
(VideoFullscreenInterfacePiP::playerController const):
(VideoFullscreenInterfacePiP::setVideoFullscreenModel):
(VideoFullscreenInterfacePiP::setVideoFullscreenChangeObserver):
(VideoFullscreenInterfacePiP::hasVideoChanged):
(VideoFullscreenInterfacePiP::videoDimensionsChanged):
(VideoFullscreenInterfacePiP::externalPlaybackChanged):
(VideoFullscreenInterfacePiP::pictureInPictureWasStartedWhenEnteringBackground const):
(VideoFullscreenInterfacePiP::applicationDidBecomeActive):
(VideoFullscreenInterfacePiP::setupFullscreen):
(VideoFullscreenInterfacePiP::enterFullscreen):
(VideoFullscreenInterfacePiP::exitFullscreen):
(VideoFullscreenInterfacePiP::cleanupFullscreen):
(VideoFullscreenInterfacePiP::invalidate):
(VideoFullscreenInterfacePiP::modelDestroyed):
(VideoFullscreenInterfacePiP::setPlayerIdentifier):
(VideoFullscreenInterfacePiP::requestHideAndExitFullscreen):
(VideoFullscreenInterfacePiP::preparedToReturnToInline):
(VideoFullscreenInterfacePiP::preparedToExitFullscreen):
(VideoFullscreenInterfacePiP::mayAutomaticallyShowVideoPictureInPicture const):
(VideoFullscreenInterfacePiP::prepareForPictureInPictureStop):
(VideoFullscreenInterfacePiP::willStartPictureInPicture):
(VideoFullscreenInterfacePiP::didStartPictureInPicture):
(VideoFullscreenInterfacePiP::failedToStartPictureInPicture):
(VideoFullscreenInterfacePiP::willStopPictureInPicture):
(VideoFullscreenInterfacePiP::didStopPictureInPicture):
(VideoFullscreenInterfacePiP::prepareForPictureInPictureStopWithCompletionHandler):
(VideoFullscreenInterfacePiP::setHasVideoContentLayer):
(VideoFullscreenInterfacePiP::setInlineRect):
(VideoFullscreenInterfacePiP::doSetup):
(VideoFullscreenInterfacePiP::preparedToReturnToStandby):
(VideoFullscreenInterfacePiP::finalizeSetup):
(VideoFullscreenInterfacePiP::doEnterFullscreen):
(VideoFullscreenInterfacePiP::doExitFullscreen):
(VideoFullscreenInterfacePiP::returnToStandby):
(VideoFullscreenInterfacePiP::setMode):
(VideoFullscreenInterfacePiP::clearMode):
(VideoFullscreenInterfacePiP::isPlayingVideoInEnhancedFullscreen const):
(WebCore::setSupportsPictureInPicture):
(WebCore::supportsPictureInPicture):
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.h: Added.
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm: Added.
(SOFT_LINK_CLASS_OPTIONAL):
(-[WebAVPlayerLayer init]):
(-[WebAVPlayerLayer dealloc]):
(-[WebAVPlayerLayer fullscreenInterface]):
(-[WebAVPlayerLayer setFullscreenInterface:]):
(-[WebAVPlayerLayer playerController]):
(-[WebAVPlayerLayer setPlayerController:]):
(-[WebAVPlayerLayer setVideoSublayer:]):
(-[WebAVPlayerLayer videoSublayer]):
(-[WebAVPlayerLayer calculateTargetVideoFrame]):
(-[WebAVPlayerLayer layoutSublayers]):
(-[WebAVPlayerLayer resolveBounds]):
(-[WebAVPlayerLayer setVideoGravity:]):
(-[WebAVPlayerLayer videoGravity]):
(-[WebAVPlayerLayer videoRect]):
(+[WebAVPlayerLayer keyPathsForValuesAffectingVideoRect]):
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(allocWebAVPictureInPicturePlayerLayerViewInstance):
(allocWebAVPlayerLayerViewInstance):
(VideoFullscreenInterfaceAVKit::videoDimensionsChanged):
(VideoFullscreenInterfaceAVKit::cleanupFullscreen):
(VideoFullscreenInterfaceAVKit::requestHideAndExitFullscreen):
(VideoFullscreenInterfaceAVKit::shouldExitFullscreenWithReason):
(VideoFullscreenInterfaceAVKit::exitFullscreenHandler):
(-[WebAVPlayerLayer init]): Deleted.
(-[WebAVPlayerLayer dealloc]): Deleted.
(-[WebAVPlayerLayer fullscreenInterface]): Deleted.
(-[WebAVPlayerLayer setFullscreenInterface:]): Deleted.
(-[WebAVPlayerLayer playerController]): Deleted.
(-[WebAVPlayerLayer setPlayerController:]): Deleted.
(-[WebAVPlayerLayer setVideoSublayer:]): Deleted.
(-[WebAVPlayerLayer videoSublayer]): Deleted.
(-[WebAVPlayerLayer calculateTargetVideoFrame]): Deleted.
(-[WebAVPlayerLayer layoutSublayers]): Deleted.
(-[WebAVPlayerLayer resolveBounds]): Deleted.
(-[WebAVPlayerLayer setVideoGravity:]): Deleted.
(-[WebAVPlayerLayer videoGravity]): Deleted.
(-[WebAVPlayerLayer videoRect]): Deleted.
(+[WebAVPlayerLayer keyPathsForValuesAffectingVideoRect]): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
(VideoFullscreenControllerContext::setUpFullscreen):
(-[WebVideoFullscreenController init]):
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::playerViewController const):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen]):
(-[WKFullScreenWindowController beganEnterFullScreenWithInitialFrame:finalFrame:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:]):
(-[WKFullScreenWindowController _dismissFullscreenViewController]):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::supportsVideoFullscreen const):
(WebKit::VideoFullscreenManager::supportsVideoFullscreenStandby const):

Canonical link: <a href="https://commits.webkit.org/252033@main">https://commits.webkit.org/252033@main</a>
</pre>

grorg commented on Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
```diff
@@ -56,18 +57,23 @@
+#if HAVE(PIP_CONTROLLER)
+SOFT_LINK_CLASS_OPTIONAL(AVKit, AVPictureInPictureController)
+SOFT_LINK_CLASS_OPTIONAL(AVKit, AVPictureInPictureControllerContentSource)
+SOFT_LINK_CLASS_OPTIONAL(AVKit, AVPictureInPictureContentViewController)
```
I'd really like to know why GitHub syntax highlights only the first macro in a set.

grorg commented on Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
```diff
@@ -98,6 +104,296 @@ @interface AVPlayerViewController (Details)
+//- (AVPlayerController*)playerController
+//{
+//    return _playerController;
+//}
```
Did you mean to leave this in? Same with another ones below. I guess you want to show what the code should be like if you didn't have to play ObjC runtime API games.

jyavenard replied:
yeah, I thought that showing the original obj-C code before it gets converted to objC runtime methods instead added values; I've removed them all

grorg commented on Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
```diff
@@ -98,6 +104,296 @@ @interface AVPlayerViewController (Details)
+static void WebAVPictureInPictureContentViewController_dealloc(id aSelf, SEL)
+{
+    WebAVPictureInPictureContentViewController *pipController = aSelf;
+    [pipController setValue:nil forKey:@"_controller"];
```
Not necessary, but maybe leave a comment saying this is necessary to ensure `CFRelease` is called.

jyavenard replied:
I've changed that code to use obj-C retain/release instead.

I believe we have an extra autorelease happening in the API, if I were to perform the logical release here, it was crash.
And unfortunately Xcode instruments to track retain/release is currently broken

grorg commented on Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
```diff
@@ -98,6 +104,296 @@ @interface AVPlayerViewController (Details)
+        return;
+    }
+
+    _startPictureInPictureTimer = [NSTimer scheduledTimerWithTimeInterval:startPictureInPictureTimeInterval repeats:NO block:^(NSTimer *_Nonnull) {
```
Any reason you use `NSTimer` rather than `WorkQueue::dispatchAfter` or just `dispatch_after`?

jyavenard replied:
that way I can cancel the timer if the KVO gets resolved quickly enough.This is to monitor if we don't get into PiP and carrying a change done by Peng with the legacy code.

grorg commented on Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
```diff
@@ -98,6 +104,296 @@ @interface AVPlayerViewController (Details)
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self->_pip startPictureInPicture];
+    });
+}
```
I don't get this at all :)

Why do you start PiP when someone observes any key path?

jyavenard replied:
> I don't get this at all :)
> 
> Why do you start PiP when someone observes any key path?

We get here if the `pictureInPicturePossible` property returned false and if so starting PiP will immediately fail. So we create a KVO observable that will wait up to xx seconds for this property to change, and only then will start PiP

grorg commented on Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
```diff
@@ -387,3 +387,32 @@ NS_ASSUME_NONNULL_BEGIN
+@interface AVPictureInPictureControllerContentSource (GenericSupport)
+- (instancetype)initWithSourceView:(UIView *)sourceView contentViewController:(AVPictureInPictureContentViewController *)contentViewController playerController:(__kindof AVPlayerController *)playerController API_AVAILABLE(ios(16.0),tvos(16.0)) API_UNAVAILABLE(macos, watchos);
+@property (nonatomic, weak, readonly, nullable) UIView * activeSourceView API_AVAILABLE(ios(16.0),tvos(16.0)) API_UNAVAILABLE(macos, watchos);
+@property (nonatomic, readonly, nullable) __kindof AVPictureInPictureContentViewController * activeContentViewController API_AVAILABLE(ios(16.0),tvos(16.0)) API_UNAVAILABLE(macos, watchos);
```
Nit: Don't we remove the space after the `*`?

jyavenard replied:
done

jyavenard committed:
Adopt new AVPictureInPictureController for handling Picture In Picture
https://bugs.webkit.org/show_bug.cgi?id=242061
rdar://96066192

Reviewed by Dean Jackson.

For now, code is only enabled if UIKIT_WEBKIT_INTERNALS is set.
While the aim is to adopt this new API across all platforms,
it is currently only usable with UIKit.

Manually tested in the simulator.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/PAL/pal/spi/cocoa/AVKitSPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerGetRawCookies const):
* Source/WebCore/platform/cocoa/VideoFullscreenInterfacePiP.h: Copied from Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h.
* Source/WebCore/platform/cocoa/VideoFullscreenInterfacePiP.mm: Added.
(clearUIColor):
(boolString):
(WebAVPictureInPictureContentViewController_initWithController):
(WebAVPictureInPictureContentViewController_controller):
(WebAVPictureInPictureContentViewController_playerLayer):
(WebAVPictureInPictureContentViewController_setPlayerLayer):
(WebAVPictureInPictureContentViewController_viewWillLayoutSubviews):
(WebAVPictureInPictureContentViewController_dealloc):
(allocWebAVPictureInPictureContentViewControllerInstance):
(-[WebAVPictureInPictureController fullscreenInterface]):
(-[WebAVPictureInPictureController setFullscreenInterface:]):
(-[WebAVPictureInPictureController initWithFullscreenInterface:]):
(-[WebAVPictureInPictureController dealloc]):
(-[WebAVPictureInPictureController tryToStartPictureInPicture]):
(-[WebAVPictureInPictureController stopPictureInPicture]):
(-[WebAVPictureInPictureController cleanup]):
(-[WebAVPictureInPictureController observeValueForKeyPath:ofObject:change:context:]):
(-[WebAVPictureInPictureController initObserver]):
(-[WebAVPictureInPictureController removeObserver]):
(-[WebAVPictureInPictureController pictureInPictureControllerWillStartPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureControllerDidStartPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureController:failedToStartPictureInPictureWithError:]):
(-[WebAVPictureInPictureController pictureInPictureControllerWillStopPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureControllerDidStopPictureInPicture:]):
(-[WebAVPictureInPictureController pictureInPictureController:restoreUserInterfaceForPictureInPictureStopWithCompletionHandler:]):
(WebAVPlayerLayerView_layerClass):
(WebAVPlayerLayerView_playerController):
(WebAVPlayerLayerView_setPlayerController):
(WebAVPlayerLayerView_playerLayer):
(allocWebAVPlayerLayerViewInstance):
(VideoFullscreenInterfacePiP::create):
(VideoFullscreenInterfacePiP::VideoFullscreenInterfacePiP):
(VideoFullscreenInterfacePiP::~VideoFullscreenInterfacePiP):
(VideoFullscreenInterfacePiP::playerController const):
(VideoFullscreenInterfacePiP::setVideoFullscreenModel):
(VideoFullscreenInterfacePiP::setVideoFullscreenChangeObserver):
(VideoFullscreenInterfacePiP::hasVideoChanged):
(VideoFullscreenInterfacePiP::videoDimensionsChanged):
(VideoFullscreenInterfacePiP::externalPlaybackChanged):
(VideoFullscreenInterfacePiP::pictureInPictureWasStartedWhenEnteringBackground const):
(VideoFullscreenInterfacePiP::applicationDidBecomeActive):
(VideoFullscreenInterfacePiP::setupFullscreen):
(VideoFullscreenInterfacePiP::enterFullscreen):
(VideoFullscreenInterfacePiP::exitFullscreen):
(VideoFullscreenInterfacePiP::cleanupFullscreen):
(VideoFullscreenInterfacePiP::invalidate):
(VideoFullscreenInterfacePiP::modelDestroyed):
(VideoFullscreenInterfacePiP::setPlayerIdentifier):
(VideoFullscreenInterfacePiP::requestHideAndExitFullscreen):
(VideoFullscreenInterfacePiP::preparedToReturnToInline):
(VideoFullscreenInterfacePiP::preparedToExitFullscreen):
(VideoFullscreenInterfacePiP::mayAutomaticallyShowVideoPictureInPicture const):
(VideoFullscreenInterfacePiP::prepareForPictureInPictureStop):
(VideoFullscreenInterfacePiP::willStartPictureInPicture):
(VideoFullscreenInterfacePiP::didStartPictureInPicture):
(VideoFullscreenInterfacePiP::failedToStartPictureInPicture):
(VideoFullscreenInterfacePiP::willStopPictureInPicture):
(VideoFullscreenInterfacePiP::didStopPictureInPicture):
(VideoFullscreenInterfacePiP::prepareForPictureInPictureStopWithCompletionHandler):
(VideoFullscreenInterfacePiP::setHasVideoContentLayer):
(VideoFullscreenInterfacePiP::setInlineRect):
(VideoFullscreenInterfacePiP::doSetup):
(VideoFullscreenInterfacePiP::preparedToReturnToStandby):
(VideoFullscreenInterfacePiP::finalizeSetup):
(VideoFullscreenInterfacePiP::doEnterFullscreen):
(VideoFullscreenInterfacePiP::doExitFullscreen):
(VideoFullscreenInterfacePiP::returnToStandby):
(VideoFullscreenInterfacePiP::setMode):
(VideoFullscreenInterfacePiP::clearMode):
(VideoFullscreenInterfacePiP::isPlayingVideoInEnhancedFullscreen const):
(WebCore::setSupportsPictureInPicture):
(WebCore::supportsPictureInPicture):
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.h: Added.
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm: Added.
(SOFT_LINK_CLASS_OPTIONAL):
(-[WebAVPlayerLayer init]):
(-[WebAVPlayerLayer dealloc]):
(-[WebAVPlayerLayer fullscreenInterface]):
(-[WebAVPlayerLayer setFullscreenInterface:]):
(-[WebAVPlayerLayer playerController]):
(-[WebAVPlayerLayer setPlayerController:]):
(-[WebAVPlayerLayer setVideoSublayer:]):
(-[WebAVPlayerLayer videoSublayer]):
(-[WebAVPlayerLayer calculateTargetVideoFrame]):
(-[WebAVPlayerLayer layoutSublayers]):
(-[WebAVPlayerLayer resolveBounds]):
(-[WebAVPlayerLayer setVideoGravity:]):
(-[WebAVPlayerLayer videoGravity]):
(-[WebAVPlayerLayer videoRect]):
(+[WebAVPlayerLayer keyPathsForValuesAffectingVideoRect]):
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(allocWebAVPictureInPicturePlayerLayerViewInstance):
(allocWebAVPlayerLayerViewInstance):
(VideoFullscreenInterfaceAVKit::videoDimensionsChanged):
(VideoFullscreenInterfaceAVKit::cleanupFullscreen):
(VideoFullscreenInterfaceAVKit::requestHideAndExitFullscreen):
(VideoFullscreenInterfaceAVKit::shouldExitFullscreenWithReason):
(VideoFullscreenInterfaceAVKit::exitFullscreenHandler):
(-[WebAVPlayerLayer init]): Deleted.
(-[WebAVPlayerLayer dealloc]): Deleted.
(-[WebAVPlayerLayer fullscreenInterface]): Deleted.
(-[WebAVPlayerLayer setFullscreenInterface:]): Deleted.
(-[WebAVPlayerLayer playerController]): Deleted.
(-[WebAVPlayerLayer setPlayerController:]): Deleted.
(-[WebAVPlayerLayer setVideoSublayer:]): Deleted.
(-[WebAVPlayerLayer videoSublayer]): Deleted.
(-[WebAVPlayerLayer calculateTargetVideoFrame]): Deleted.
(-[WebAVPlayerLayer layoutSublayers]): Deleted.
(-[WebAVPlayerLayer resolveBounds]): Deleted.
(-[WebAVPlayerLayer setVideoGravity:]): Deleted.
(-[WebAVPlayerLayer videoGravity]): Deleted.
(-[WebAVPlayerLayer videoRect]): Deleted.
(+[WebAVPlayerLayer keyPathsForValuesAffectingVideoRect]): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
(VideoFullscreenControllerContext::setUpFullscreen):
(-[WebVideoFullscreenController init]):
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::playerViewController const):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen]):
(-[WKFullScreenWindowController beganEnterFullScreenWithInitialFrame:finalFrame:]):
(-[WKFullScreenWindowController beganExitFullScreenWithInitialFrame:finalFrame:]):
(-[WKFullScreenWindowController _dismissFullscreenViewController]):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::VideoFullscreenManager::supportsVideoFullscreen const):
(WebKit::VideoFullscreenManager::supportsVideoFullscreenStandby const):

Canonical link: https://commits.webkit.org/252033@main

cdumez commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -650,6 +650,10 @@
 #endif
 
+#if HAVE(UIKIT_WEBKIT_INTERNALS)
+#define HAVE_PIP_CONTROLLER 1
```
This should be in PlatformHave, not PlatformEnable since it has "HAVE" in the name.
